src/onegov/org/views/resource.py

```diff
@@ -969,26 +969,42 @@ def view_resource(
 def handle_delete_resource(self: Resource, request: OrgRequest) -> None:
 
     request.assert_valid_csrf_token()
-    tickets = TicketCollection(request.session)
 
-    def handle_reservation_tickets(reservation: BaseReservation) -> None:
-        ticket = tickets.by_handler_id(reservation.token.hex)
+    def handle_reservation_ticket(ticket: ReservationTicket) -> None:
         if ticket:
             assert request.current_user is not None
 
             close_ticket(ticket, request.current_user, request)
             ticket.create_snapshot(request)
 
-            payment = ticket.handler.payment
-            if (payment and PaymentCollection(request.session).query()
-                    .filter_by(id=payment.id).first()):
-                PaymentCollection(request.session).delete(payment)
+            # unlink payment from invoice items, delete invoice
+            # items and finally delete invoice and unlink from ticket
+            if ticket.invoice:
+                for invoice_item in ticket.invoice.items:
+                    invoice_item.payments = []
+                    request.session.delete(invoice_item)
+
+                request.session.delete(ticket.invoice)
+
+                # unlink invoice from ticket
+                ticket.invoice = None
+                ticket.invoice_id = None
+
+            # unlink payment from reservation
+            for reservation in ticket.handler.reservations:
+                reservation.payment = None
+
+            if ticket.payment:
+                # delete payment from ticket
+                request.session.delete(ticket.payment)
+                ticket.payment = None
+                ticket.payment_id = None
```

Comment:
```suggestion
            if ticket.payment:
              # unlink payment from reservation
              for reservation in ticket.handler.reservations:
                  reservation.payment = None

                # delete payment from ticket
                request.session.delete(ticket.payment)
                ticket.payment = None
                ticket.payment_id = None
```

This way we only spend the cost of looking up the reservations if there actually is a payment.